BUG: improve precision of geomspace

ashwinvis opened the pull request:
Previously `numpy.core.function_base.geomspace` relied on a `logspace`
function call with `base=10`. This provided accurate results for
geometric progressions with ratio=10, but compromised accuracy for all
other progressions. This commit tries to accurately compute the common
ratio and use that the base instead. `decimal.Decimal` type is used to
improve the precision for real geom. progressions. This allows one to
use `geomspace` with `dtype=int` argument. Fixes #9987.

ashwinvis committed:
BUG: improve precision of geomspace

 - Stricter tests for dtype=int
 - Add known failure test with reference to gh-9993
 - Support for special cases of geomspace and use _nx.log
 - Special casing geomspace to calculate sequences with ratio 2 or 10

Previously `numpy.core.function_base.geomspace` relied on a `logspace`
function call with `base=10`. This provided accurate results for
geometric progressions with ratio=10, but compromised accuracy for all
other progressions. This commit tries to accurately compute the common
ratio and use that the base instead. `decimal.Decimal` type is used to
improve the precision for real geom. progressions. This allows one to
use `geomspace` with `dtype=int` argument. Fixes #9987.

Added support for function calls like `geomspace(x, x**y, num=arbitary)`
for which base=x will be used. Also `log2` and `log10` will be used only
when necessary, and use `log` most of the time.

Geomspace calls logspace with base 2 or 10 irrespective of value num if
suitable for function calls like `geomspace(1, base**y, num=arbitary)`.

endolith commented on numpy/core/tests/test_function_base.py:
```diff
@@ -82,14 +83,14 @@ def test_basic(self):
         assert_(len(y) == 50)
         y = geomspace(1, 1e6, num=100)
-        assert_(y[-1] == 10 ** 6)
+        assert_almost_equal(y[-1], 10 ** 6)
```
Why are you relaxing this test?

ashwinvis replied:
See my reply below.

ashwinvis replied:
I have reverted the test.

ashwinvis commented:
For that particular test, the common ratio is not an integer. With the earlier implementation, with base=10, the endpoint was exactly matched for this test, because logspace gets whole numbers as arguments. Now it gets floating points. But endpoint not matching is another issue with geomspace, that existed for all other cases.

-------- Original Message --------
On 10 Nov 2017, 01:42, endolith wrote:

> @endolith commented on this pull request.
>
> ---------------------------------------------------------------
>
> In [numpy/core/tests/test_function_base.py](https://github.com/numpy/numpy/pull/9993#discussion_r150128901):
>
>> @@ -82,14 +83,14 @@ def test_basic(self):
>          y = geomspace(1, 1e6)
>          assert_(len(y) == 50)
>          y = geomspace(1, 1e6, num=100)
> -        assert_(y[-1] == 10 ** 6)
> +        assert_almost_equal(y[-1], 10 ** 6)
>
> Why are you relaxing this test?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, [view it on GitHub](https://github.com/numpy/numpy/pull/9993#pullrequestreview-75635358), or [mute the thread](https://github.com/notifications/unsubscribe-auth/AIuyJxbJgnTai9_iyjsAjFcryel_Nj7Gks5s05wRgaJpZM4QYlkK).

eric-wieser commented:
>  But endpoint not matching is another issue with geomspace, that existed for all other cases.

Note that this is fixable with an implementation like:

```
def geomspace(a, b, num):
	frac = np.linspace(0, 1, num)
	return a**(1-frac) * b**(frac)
```

Although that still fails on other cases.

ashwinvis commented:
@eric-wieser True! But just the endpoint business. I just tried your approach:

```
>>> np.set_printoptions(precision=16)
>>>  print(geomspace(3, 81, 4))
[ 3.                 8.999999999999998 26.999999999999996 81.               ]
```

The real culprit in either implementations is operations like `x  ** y` where y is a fraction a.k.a. finding n-th roots. A feasible way to get around is use of `decimal.Decimal` types - however that is not compatible with `numpy` types, I think.

ashwinvis commented on numpy/core/tests/test_function_base.py:
```diff
@@ -82,14 +83,14 @@ def test_basic(self):
 
         y = geomspace(8, 2, num=3)
-        assert_allclose(y, [8, 4, 2])
+        assert_array_equal(y, [8, 4, 2])
```
I hope strengthening this test balances the scales 😃

eric-wieser commented on numpy/core/function_base.py:
```diff
@@ -350,9 +395,25 @@ def geomspace(start, stop, num=50, endpoint=True, dtype=None):
+
+        log_base = _nx.log2(base)
+        log_start = _nx.log2(start) / log_base
+        log_stop = _nx.log2(stop) / log_base
```
Special casing `base == 10` here would keep precision for the base 10 cases too,

mhvk replied:
Might things be slightly more precise by using `_nx.log` above? Now, internally there are likely a number of divisions by `log2`

ashwinvis replied:
The test
```
y = geomspace(1, 1e6, num=7)
array_equal(y, [1, 10, 100, 1e3, 1e4, 1e5, 1e6])
```
proves that `base == 10` works. What @endolith mentioned can only be fixed by special casing whenever `start` and `stop` are powers of 10, irrespective of the common ratio, but do we really need that?

ashwinvis replied:
@mhvk That was what I tried in the beginning. For some reason then, the test `assert_array_equal(y, [1, 10, 100, 1e3, 1e4, 1e5, 1e6])` fails, because:
```python
>>> print(y)
[1.000000000000000e+00 9.999999999999998e+00 9.999999999999994e+01
 9.999999999999990e+02 9.999999999999989e+03 9.999999999999980e+04
 9.999999999999979e+05]
```

With `log2` we get exactly `[1.e+00 1.e+01 1.e+02 1.e+03 1.e+04 1.e+05 1.e+06]`. I don't know why honestly! Must be because `log` has an a irrational base.

ashwinvis replied:
Using `log` mostly now. Special cases for `log2` and `log10` added.

eric-wieser commented on numpy/core/tests/test_function_base.py:
```diff
@@ -149,6 +150,14 @@ def test_dtype(self):
         # Native types
+        for x in range(1, 10):
+            y = geomspace(x, x**4, num=4, dtype=int)
+            assert_array_equal(y, [x, x**2, x**3, x**4])
```
Does `num=9`, `assert_(y[::2] == [x, x**2, x**3, x**4])` work?

ashwinvis replied:
I am confused. Did you mean
```
y = geomspace(x, x**9, num=9, dtype=int)
assert_(y[::2] == [x, x**3, x**5, x**7, x**9])
```

eric-wieser replied:
Sorry, I can't count. I mean
```
geomspace(x, x**4, num=7, dtype=int)
assert_(y[::2] == [x, x**2, x**3, x**4])
```

ashwinvis replied:
Sadly it doesn't! 😞 That is a very good test case. What happens to this PR now then?

ashwinvis replied:
IMHO, while it is possible to detect `x` as the `base` in your test case, simply by checking:
```python
mod(log(stop) / log(start), 1) == 0  # is integer
``` 
However, it is next to impossible to detect `x` when `start=1` and `num` is arbitary. i.e. for function calls like `geomspace(1, 10 ** 6, num=100, dtype=int)` or `geomspace(1, 3 ** 6, num=27, dtype=int)` etc.

mhvk commented:
It doesn't solve the issue, but I wonder if it wouldn't at least be clearer to just use a simpler implementation:
```
start*(stop/start)**np.linspace(0, 1, num, endpoint=endpoint)
```
It at least allows one to get rid of some of the complex arrangements for ensuring `log` doesn't fail.

It may also make sense to special case powers of integers, as they are common.

ashwinvis commented:
In retrospect I squashed the commits too soon, apologies. But to summarize `geomspace` can now handle:
* `geomspace(x, x**n, num=any, dtype=int)`
* `geomspace(1, 10**n, num=any, dtype=int)`
* `geomspace(1, 2**n, num=any, dtype=int)`

and other variants. Tests have been added for such function calls. Now geomspace relies mostly on `_nx.log`.

### Caveats
* `geomspace(1, x**n, num=any, dtype=int)` when `x is not in [2, 10]`. Added known failure test for this.
* `geomspace(x, x**n, num=any)` matching endpoint exactly is still an issue when `x is not in [2, 10]`
